### PR TITLE
types: fix STR_TO_DATE handling for %h, %r (#18171) (#18428)

### LINKED ISSUE #18171: StrToDate handling of %h(12) is not as expected
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
SELECT STR_TO_DATE('01,5,2013 12:30:17','%d,%m,%Y %h:%i:%s');

### 2. What did you expect to see? (Required)
mysql 8.0.17
2013-05-01 00:30:17
### 3. What did you see instead (Required)
2013-05-01 12:30:17
### 4. Affected version (Required)

v4.0.0

### PR DESCRIPTION
**Pull request by ti-srebot**
cherry-pick #18428 to release-3.0

---

### What problem does this PR solve?

Issue Number: close #18171 

Problem Summary: Behavior of function `STR_TO_DATE` is not compatible with MySQL.
- '**%h**:%i:%s' parses '12:mm:ss' into '12:mm:ss', which should be '00:mm:ss'.
- '**%r**' parses '12:mm:ss AM' into '12:mm:ss' and parses '12:hh:ss PM' into 'NULL'.

### What is changed and how it works?

What's Changed:
- The hour part will change to 0 if '%h' matches 12 and no AM or PM occurs.
- The hour part will subtract 12 if '%r' matches pattern '12:mm:ss AM' or '12:mm:ss PM'.

How it Works:
- Function `mysqlTimeFix` inspects the occurrence of '%h', '%p' and whether the hour part is 12.
- Function `time12Hour` now fixes the time when the hour part is 12.

Tests 

- Unit test
- Manual test

Side effects

No.

### Release note 

- types: fix STR_TO_DATE's handling for format token '%r', '%h'

### CODE REVIEW

File: types/format_test.go [hunk 1]
```diff
@@ -88,10 +88,16 @@ func (s *testTimeSuite) TestStrToDate(c *C) {
 		{`May 01, 2013`, `%M %d,%Y`, types.FromDate(2013, 5, 1, 0, 0, 0, 0)},
 		{`a09:30:17`, `a%h:%i:%s`, types.FromDate(0, 0, 0, 9, 30, 17, 0)},
 		{`09:30:17a`, `%h:%i:%s`, types.FromDate(0, 0, 0, 9, 30, 17, 0)},
+<<<<<<< HEAD
 		{`abc`, `abc`, types.ZeroTime},
+=======
+		{`12:43:24`, `%h:%i:%s`, types.FromDate(0, 0, 0, 0, 43, 24, 0)},
+		{`abc`, `abc`, types.ZeroCoreTime},
+>>>>>>> cc0b6f1... types: fix STR_TO_DATE handling for %h, %r (#18171) (#18428)
```
Comment (qw4990) [inline types/format_test.go:96]: Please resolve conflicts. @dyzsr

Comment (qw4990) [inline types/format_test.go:96]: Please resolve conflicts. @dyzsr

**Comment by XuHuaiyu** [comment]
@qw4990 can merge now

**Comment by qw4990** [comment]
> @qw4990 can merge now

I have no permission, PTAL @zz-jason @SunRunAway

**Pull request merged**